File: app/src/main/java/com/pr0gramm/app/ui/views/viewer/GifMediaView.java
package com.pr0gramm.app.ui.views.viewer;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.view.View;
import android.widget.ImageView;

import com.google.common.base.Optional;
import com.jakewharton.rxbinding.view.RxView;
import com.pr0gramm.app.ActivityComponent;
import com.pr0gramm.app.R;
import com.pr0gramm.app.Settings;
import com.pr0gramm.app.services.GifDrawableLoader;
import com.pr0gramm.app.ui.views.BusyIndicator;

import javax.inject.Inject;

import butterknife.Bind;
import pl.droidsonroids.gif.GifDrawable;

import static com.pr0gramm.app.ui.dialogs.ErrorDialogFragment.defaultOnError;
import static com.pr0gramm.app.util.AndroidUtility.checkMainThread;
import static com.pr0gramm.app.util.AndroidUtility.endAction;

/**
 */
@SuppressLint("ViewConstructor")
public class GifMediaView extends AbstractProgressMediaView {
    @Inject
    Settings settings;

    @Inject
    GifDrawableLoader gifDrawableLoader;

    @Bind(R.id.image)
    ImageView imageView;

    // the gif that is shown
    private GifDrawable gif;

    public GifMediaView(Activity context, MediaUri url, Runnable onViewListener) {
        super(context, R.layout.player_gif, url.withProxy(true), onViewListener);
        imageView.setAlpha(0.f);
        loadGif();

        // cleanup on detach!
        RxView.detaches(this).subscribe(event -> {
            imageView.setImageDrawable(null);

            if (gif != null) {
                gif.recycle();
                gif = null;
            }
        });
    }

    private void loadGif() {
        showBusyIndicator();

        gifDrawableLoader.load(getEffectiveUri())
                .compose(backgroundBindView())
                .doAfterTerminate(this::hideBusyIndicator)
                .subscribe(this::onDownloadStatus, defaultOnError());
    }

    private void onDownloadStatus(GifDrawableLoader.DownloadStatus state) {
        checkMainThread();

        onDownloadProgress(state.progress);

        if (state.finished()) {
            gif = state.drawable;
            imageView.setImageDrawable(this.gif);

            setViewAspect((float) gif.getIntrinsicWidth() / gif.getIntrinsicHeight());

            if (isPlaying()) {
                imageView.animate().alpha(1.f)
                        .setListener(endAction(this::onMediaShown))
                        .setDuration(ANIMATION_DURATION)
                        .start();
            } else {
                imageView.setAlpha(1.f);
                gif.stop();
            }
        }
    }

    private void onDownloadProgress(float progress) {
        checkMainThread();

        View progressView = getProgressView();
        if (progressView instanceof BusyIndicator) {
            BusyIndicator bar = (BusyIndicator) progressView;
            bar.setProgress(progress);
        }
    }

    @Override
    protected void onPreviewRemoved() {
        imageView.setVisibility(VISIBLE);
    }

    @Override
    public void onResume() {
        super.onResume();
        if (gif != null && isPlaying()) {
            gif.start();
            onMediaShown();
        }
    }

    @Override
    protected Optional<Float> getVideoProgress() {
        if (gif != null && isPlaying()) {
            int position = gif.getCurrentFrameIndex();
            int duration = gif.getNumberOfFrames();

            if (position >= 0 && duration > 0) {
                return Optional.of(position / (float) duration);
            }
        }

        return Optional.absent();
    }

    @Override
    protected void injectComponent(ActivityComponent component) {
        component.inject(this);
    }

    @Override
    public void onPause() {
        super.onPause();
        if (gif != null && isPlaying())
            gif.pause();
    }

    @Override
    public void playMedia() {
        super.playMedia();
        if (gif != null && isPlaying()) {
            gif.start();
            onMediaShown();
        }
    }

    @Override
    public void stopMedia() {
        super.stopMedia();
        if (gif != null)
            gif.stop();
    }

    @Override
    public void rewind() {
        if (gif != null && isPlaying()) {
            gif.seekTo(0);
        }
    }
}

File: app/src/main/java/com/pr0gramm/app/ui/views/viewer/GifMediaView.kt
package com.pr0gramm.app.ui.views.viewer

import android.annotation.SuppressLint
import androidx.core.view.isVisible
import com.pr0gramm.app.Duration
import com.pr0gramm.app.R
import com.pr0gramm.app.databinding.PlayerGifBinding
import com.pr0gramm.app.services.GifDrawableLoader
import com.pr0gramm.app.ui.base.onAttachedScope
import com.pr0gramm.app.ui.views.BusyIndicator
import com.pr0gramm.app.ui.views.instance
import com.pr0gramm.app.util.addOnDetachListener
import com.pr0gramm.app.util.checkMainThread
import pl.droidsonroids.gif.GifDrawable

/**
 */
@SuppressLint("ViewConstructor")
class GifMediaView(config: Config) : AbstractProgressMediaView(config, R.layout.player_gif) {
    private val gifDrawableLoader: GifDrawableLoader by instance()

    private val views = PlayerGifBinding.bind(this)

    // the gif that is shown
    private var gif: GifDrawable? = null

    init {
        views.image.alpha = 0f

        onAttachedScope {
            if (gif == null) {
                loadGif()
            }
        }

        // cleanup on detach!
        addOnDetachListener {
            views.image.setImageDrawable(null)

            // recycle if possible
            gif?.recycle()
            gif = null
        }
    }

    private suspend fun loadGif() {
        showBusyIndicator()

        try {
            gifDrawableLoader.load(effectiveUri).collect { state ->
                onDownloadStatus(state)
            }
        } finally {
            hideBusyIndicator()
        }
    }

    private fun onDownloadStatus(state: GifDrawableLoader.State) {
        checkMainThread()

        onDownloadProgress(state.progress)

        if (state.drawable != null && isAttachedToWindow) {
            this.gif = state.drawable

            views.image.setImageDrawable(state.drawable)

            viewAspect = state.drawable.intrinsicWidth.toFloat() / state.drawable.intrinsicHeight

            if (isPlaying) {
                gif?.start()

                views.image.animate().alpha(1f)
                    .withEndAction { onMediaShown() }
                    .setDuration(MediaView.ANIMATION_DURATION)
                    .start()
            } else {
                views.image.alpha = 1f
                state.drawable.stop()
            }
        }
    }

    private fun onDownloadProgress(progress: Float) {
        checkMainThread()
        (busyIndicator as? BusyIndicator)?.progress = progress
    }

    override fun onPreviewRemoved() {
        views.image.isVisible = true
    }

    override fun currentVideoProgress(): ProgressInfo? {
        val gif = gif

        if (isPlaying && gif != null) {
            val position = gif.currentFrameIndex
            val duration = gif.numberOfFrames

            if (position >= 0 && duration > 0) {
                return ProgressInfo(
                    position / duration.toFloat(), 1f,
                    duration = Duration.millis(gif.duration.toLong())
                )
            }
        }

        return null
    }

    override fun playMedia() {
        super.playMedia()

        gif.takeIf { isPlaying }?.let {
            gif?.start()
            onMediaShown()
        }
    }

    override fun stopMedia() {
        super.stopMedia()
        gif?.stop()
    }

    override fun rewind() {
        this.gif?.seekTo(0)
    }
}


